File: accounts/modules/addons/ms365backup/lib/Ms365Backup/CalendarGraphFields.php
<?php
declare(strict_types=1);

namespace Ms365Backup;

final class CalendarGraphFields
{
    public const LIST_SELECT = 'id,createdDateTime,lastModifiedDateTime,changeKey,iCalUId,type,seriesMasterId,recurrence,start,end,originalStart,isCancelled,hasAttachments,attendees,organizer,body,locations,onlineMeeting';

    public const SERIES_GET_SELECT = 'id,createdDateTime,lastModifiedDateTime,changeKey,iCalUId,type,recurrence,start,end,cancelledOccurrences,exceptionOccurrences';

    /** @var array<string, string> */
    public const PREFER_IMMUTABLE = ['Prefer' => 'IdType="ImmutableId"'];

    public const NORMAL_PAGE_SIZE = '100';

    public const PARTITION_PAGE_SIZE = '25';

    public const INVENTORY_START = '1990-01-01T00:00:00Z';

    /** OData $filter for Edm.DateTimeOffset — literals must not be quoted (Graph 400 otherwise). */
    public static function createdDateTimeFilter(\DateTimeImmutable $start, \DateTimeImmutable $end): string
    {
        return sprintf(
            'createdDateTime ge %s and createdDateTime lt %s',
            $start->format('Y-m-d\TH:i:s\Z'),
            $end->format('Y-m-d\TH:i:s\Z'),
        );
    }
}

File: ms365-backup-worker/internal/graphsync/calendar_graph_fields.go
package graphsync

// CalendarGraphFields mirrors PHP Ms365Backup\CalendarGraphFields.
const (
	CalendarListSelect     = "id,createdDateTime,lastModifiedDateTime,changeKey,iCalUId,type,seriesMasterId,recurrence,start,end,originalStart,isCancelled,hasAttachments,attendees,organizer,body,locations,onlineMeeting"
	CalendarSeriesSelect   = "id,createdDateTime,lastModifiedDateTime,changeKey,iCalUId,type,recurrence,start,end,cancelledOccurrences,exceptionOccurrences"
	CalendarPreferImmutable = `IdType="ImmutableId"`
	CalendarNormalPageSize = "100"
	CalendarPartitionPageSize = "25"
	CalendarInventoryStart = "1990-01-01T00:00:00Z"
)

// CalendarImmutableHeaders returns Prefer header for stable event IDs.
func CalendarImmutableHeaders() map[string]string {
	return map[string]string{"Prefer": CalendarPreferImmutable}
}

// CreatedDateTimeFilter builds an OData filter with unquoted datetime literals.
func CreatedDateTimeFilter(start, end string) string {
	return "createdDateTime ge " + start + " and createdDateTime lt " + end
}

// LastModifiedWatermarkFilter builds incremental inventory filter.
func LastModifiedWatermarkFilter(watermark string) string {
	return "lastModifiedDateTime ge " + watermark
}
